near-sdk/src/types/account_id.rs

```diff
@@ -0,0 +1,200 @@
+    Serialize,
+    Hash,
+    BorshSchema,
+)]
+pub struct AccountId(String);
+
+impl AccountId {
+    /// Returns reference to the account ID bytes.
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+    /// Returns reference to the account ID string.
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+    /// Consumes `self` to return inner [`String`] representation.
+    pub fn into_string(self) -> String {
+        self.0
+    }
+    /// Constructs new AccountId from `String` without checking validity.
+    pub fn new_unchecked(id: String) -> Self {
+        Self(id)
+    }
+}
+
+impl fmt::Display for AccountId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<AccountId> for String {
+    fn from(id: AccountId) -> Self {
+        id.0
+    }
+}
+
+impl AsRef<str> for AccountId {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl<'de> Deserialize<'de> for AccountId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, <D as de::Deserializer<'de>>::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let s: String = Deserialize::deserialize(deserializer)?;
+        Self::try_from(s).map_err(de::Error::custom)
+    }
+}
+
+impl TryFrom<&str> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+impl TryFrom<&[u8]> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        core::str::from_utf8(value)
+            .map_err(|_| ParseAccountIdError { kind: ParseAccountIdErrorKind::InvalidUtf8 })?
+            .parse()
+    }
+}
+
+fn validate_account_id(id: &str) -> Result<(), ParseAccountIdError> {
+    if is_valid_account_id(id.as_bytes()) {
+        Ok(())
+    } else {
+        Err(ParseAccountIdError { kind: ParseAccountIdErrorKind::InvalidAccountId })
+    }
+}
+
+impl TryFrom<Vec<u8>> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        Self::try_from(
+            String::from_utf8(value)
+                .map_err(|_| ParseAccountIdError { kind: ParseAccountIdErrorKind::InvalidUtf8 })?,
+        )
+    }
+}
+
+impl TryFrom<String> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        validate_account_id(value.as_str())?;
+        Ok(Self(value))
+    }
+}
+
+impl std::str::FromStr for AccountId {
+    type Err = ParseAccountIdError;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        validate_account_id(value)?;
+        Ok(Self(value.to_string()))
+    }
+}
+
+#[derive(Debug)]
+pub struct ParseAccountIdError {
+    kind: ParseAccountIdErrorKind,
+}
+
+#[derive(Debug)]
+enum ParseAccountIdErrorKind {
+    InvalidAccountId,
+    InvalidUtf8,
+}
```

Comment:
Let's derive at least `Clone` as well, as that's impossible to provide at call-site. We might also derive PartialEq/Eq